pkg/kubelet/pluginmanager/plugin_manager.go

```diff
@@ -0,0 +1,132 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pluginmanager
+
+import (
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/klog"
+	"k8s.io/kubernetes/pkg/kubelet/config"
+	"k8s.io/kubernetes/pkg/kubelet/pluginmanager/cache"
+	"k8s.io/kubernetes/pkg/kubelet/pluginmanager/metrics"
+	"k8s.io/kubernetes/pkg/kubelet/pluginmanager/operationexecutor"
+	"k8s.io/kubernetes/pkg/kubelet/pluginmanager/pluginwatcher"
+	"k8s.io/kubernetes/pkg/kubelet/pluginmanager/reconciler"
+)
+
+// PluginManager runs a set of asynchronous loops that figure out which plugins
+// need to be registered/deregistered and makes it so.
+type PluginManager interface {
+	// Starts the plugin manager and all the asynchronous loops that it controls
+	Run(sourcesReady config.SourcesReady, stopCh <-chan struct{})
+
+	// AddHandler adds the given plugin handler for a specific plugin type, which
+	// will be added to the actual state of world cache so that it can be passed to
+	// the desired state of world cache in order to be used during plugin
+	// registration/deregistration
+	AddHandler(pluginType string, pluginHandler cache.PluginHandler)
+
+	// GetRegisteredPlugins generates and returns a list of plugins that are
+	// successfully registered plugins in the current actual state of world.
+	GetRegisteredPlugins() []cache.PluginInfo
```

Comment:
This method is only used by `plugin_manager_test.go`. That indicates the unit test is making assumptions about implementation it should not.

To fix this, let's do this:
1. Remove `GetRegisteredPlugins()` from this interface.
2. Create a "mock" or "fake" `pluginHandler cache.PluginHandler` that keeps track of when (and maybe how many times) "Validate", "Register" or "Unregister" are called for each socket, and the fake should expose methods that allow the tester to inspect that at some point after construction and execution. See https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/volumemanager/volume_manager_fake.go for a rough example of this.
3. Modify `plugin_manager_test.go` so that instead of using `GetRegisteredPlugins()` to figure out if a plugin is registered, it uses the "fake" or "mock" you created above.

I know this is a non-trivial amount of work, and if you do not want to do it now, make sure to open an issue to track the work so it gets addressed in a follow up.